app/api/business/team_business.py

```diff
@@ -0,0 +1,309 @@
+    return people_overview
+
+
+def get_team(team_id):
+    team = team_service.get_team(team_id)
+    if not team:
+        raise NotFoundError('Team {} does not exist'.format(team_id))
+
+    team_member = team_member_service.find(
+        is_team_lead=True,
+        team_id=team_id,
+        user_id=current_user.id
+    ).one_or_none()
+
+    if not team_member:
+        raise UnauthorisedError('Only team leads can edit a team')
+
+    domain = get_email_domain(current_user.email_address)
+
+    if team['teamMembers'] is not None:
+        for user_id, team_member in team['teamMembers'].iteritems():
+            missing_permissions = [permission for permission in permission_types
+                                   if permission not in team_member['permissions']]
+
+            for permission in missing_permissions:
+                team_member['permissions'][permission] = False
+
+    team.update(domain=domain)
+
+    return team
+
+
+def update_team(team_id, data):
+    team = team_service.get_team_for_update(team_id)
+    if not team:
+        raise NotFoundError('Team {} does not exist'.format(team_id))
+
+    team_member = team_member_service.find(
+        is_team_lead=True,
+        team_id=team_id,
+        user_id=current_user.id
+    ).one_or_none()
+    if not team_member:
+        raise UnauthorisedError('Only team leads can edit a team')
+
+    update_team_information(team, data)
+    result = update_team_leads_and_members(team, data)
+    update_permissions(team, data)
+
+    create_team = data.get('createTeam', False)
+    saved = False
+    if create_team:
+        validation_result = TeamValidator(team, current_user).validate_all()
+        if len([e for e in validation_result.errors]) > 0:
+            raise ValidationError([e for e in validation_result.errors])
+
+        send_team_lead_notification_emails(team_id)
+        send_team_member_notification_emails(team_id)
+
+        team.status = 'completed'
+        team_service.save(team)
+        saved = True
+    elif team.status == 'completed':
+        validation_result = TeamValidator(team, current_user).validate_all()
+        if len([e for e in validation_result.errors]) > 0:
+            raise ValidationError([e for e in validation_result.errors])
+
+        team_service.save(team)
+        saved = True
+        new_team_leads = result.get('new_team_leads', [])
+        new_team_members = result.get('new_team_members', [])
+        removed_team_members = result.get('removed_team_members', [])
+
+        if len(new_team_leads) > 0:
+            send_team_lead_notification_emails(team_id, new_team_leads)
+        if len(new_team_members) > 0:
+            send_team_member_notification_emails(team_id, new_team_members)
+        if len(removed_team_members) > 0:
+            send_removed_team_member_notification_emails(team_id, removed_team_members)
+    elif team.status == 'created':
+        validation_result = TeamValidator(team, current_user).validate_team_creator()
```

Comment:
I think we may also need to add a `. validate_team_members()` and `.validate_team_creator()` here, at least when the request contains data in these fields, because similar to how I was able to iterate the user base for names and email addresses when editing an already created team, I'm able to do the same for a not yet created/draft team. Basically, I can add any user id I want to either the `teamLeads` or `teamMembers` objects in the request, and the response will enrich the id with a email address and, in the case of team members, their full name as well.

Reply:
updated